#160056373 Users should be able to recieve and use a reset password link

mppdg opened the pull request:
#### What does this PR do?
- This PR will enable a user to receive a reset password link and update his/her password.

#### Description of Task to be completed?
- provide forget password links to be added to`/ login` and `/signup` page
- create a page `/forgot-password` where a user can enter the email to receive the reset password link
- ensure user receives an email
- redirect a user to `/reset-password` page with the form to update a password.
- update a user password when he provides a valid password
- show the error messages when a password is not valid or not updated

#### How should this be manually tested?
- clone repo and install dependences with `npm install`
- `npm run dev` to run locally
- navigate to `/forgot-password` to enter an email and send link
- check email and click on the reset password link
- enter the new password and on the provided form

#### What are the relevant pivotal tracker stories?
#160056373

![image](https://user-images.githubusercontent.com/39072402/45635826-f18d4800-ba9d-11e8-9d21-7b4d362c6832.png)
<img width="418" alt="screen shot 2018-09-17 at 5 17 55 pm" src="https://user-images.githubusercontent.com/39072402/45636405-893f6600-ba9f-11e8-8651-940bc0553edd.png">
<img width="419" alt="screen shot 2018-09-17 at 5 17 04 pm" src="https://user-images.githubusercontent.com/39072402/45636406-893f6600-ba9f-11e8-9d43-1780477597d6.png">
<img width="416" alt="screen shot 2018-09-17 at 5 15 01 pm" src="https://user-images.githubusercontent.com/39072402/45636407-893f6600-ba9f-11e8-90c3-6eafc1bf8c90.png">
<img width="414" alt="screen shot 2018-09-17 at 5 14 28 pm" src="https://user-images.githubusercontent.com/39072402/45636408-89d7fc80-ba9f-11e8-819a-c70a37c99d9d.png">
<img width="424" alt="screen shot 2018-09-17 at 5 13 38 pm" src="https://user-images.githubusercontent.com/39072402/45636410-89d7fc80-ba9f-11e8-9e14-afa3a27dc082.png">
<img width="583" alt="screen shot 2018-09-17 at 5 12 46 pm" src="https://user-images.githubusercontent.com/39072402/45636411-89d7fc80-ba9f-11e8-9443-1e318947d240.png">

JCanaks commented on public/styles/App.scss:
```diff
@@ -0,0 +1,14 @@
+$body-color : green;
```
Pls refactor to follow the folder structure and convention styles

JCanaks commented on public/styles/ResetPassword.scss:
```diff
@@ -0,0 +1,67 @@
+@import url("https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css");
```
Pls refactor to follow the folder structure and convention styles

JCanaks commented on src/views/user/ForgotPassword.js:
```diff
@@ -0,0 +1,120 @@
+import React from 'react';
```
Don't forget to covert all your component files to jsx

tutugodfrey committed:
chore(token-header): add token header (#20)

added token to axios instannce

[Finishes #160552899]

megame24 commented on src/configs/routes.js:
```diff
@@ -12,104 +12,103 @@ const {
 
 export default [
   {
-    id: 1,
```
Why did you remove the `id`s

megame24 commented on src/services/requestHandler.js:
```diff
@@ -0,0 +1,37 @@
+ * it for making API request
+ * @returns { instance } an instance of axios
+ */
+// const axiosInstance = () => {
```
No commented blocks of code

megame24 commented on src/services/requestHandler.js:
```diff
@@ -0,0 +1,37 @@
+// };
+
+
+const axiosInstance = (token = '') => {
```
There is an updated `axiosInstance`, you can rebase and make use of it.

megame24 commented on src/views/user/ResetPassword/ResetPassword.jsx:
```diff
@@ -0,0 +1,306 @@
+  componentDidMount() {
+    const { location, dispatch, history } = this.props;
+    const resetToken = location.search.substr(7);
+    axiosInstance().get(`/users/reset-password?token=${resetToken}`)
```
You can make use of any of the middleware in place to handle this `async call`

megame24 commented:
Handle `PropTypes validation` in your tests

mppdg committed:
Merge branch 'staging' of https://github.com/andela/rivendell-ah-client into staging

rhenshaw56 commented on public/styles/ResetPassword.scss:
```diff
@@ -0,0 +1,131 @@
+@import url("https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css");
```
rename style file to have a kebab case, sorta like `reset-password.scss`

rhenshaw56 commented on src/actions/resetPasswordActions.js:
```diff
@@ -0,0 +1,47 @@
+);
+
+const sendResetPasswordLink = (path, email) => {
+  dispatch(resetPasswordLink(path, email));
```
I had previously said not to use dispatch here since it seems sorta redundant to still call that sae action in side the same file

mppdg replied:
I have made the adjustment

rhenshaw56 commented on src/actions/resetPasswordActions.js:
```diff
@@ -0,0 +1,47 @@
+};
+
+const updatePassword = (path, password) => {
+  dispatch(resetPassword(path, password));
```
see comment above

rhenshaw56 commented on src/actions/resetPasswordActions.js:
```diff
@@ -0,0 +1,47 @@
+    type: RESET_PASSWORD,
+    payload: axiosInstance().put(
+      path,
+      { user: { password, confirm: password } },
```
I had previously stated to somehow combine reset password actions into the auth actions or user actions, same goes with the reducers

mppdg replied:
ok

rhenshaw56 commented on src/reducers/resetPasswordReducers.js:
```diff
@@ -0,0 +1,119 @@
+
+const resetPassword = (state = prevState, action = {}) => {
+  switch (action.type) {
+  case `${RESET_PASSWORD}_TOKEN`:
```
why don't you append the `_TOKEN` suffix in action types already since I know the promise middleware won't be doing that for you, except its a success, loading or failure

rhenshaw56 commented on src/views/user/ResetPassword/ForgotPassword.jsx:
```diff
@@ -0,0 +1,153 @@
+   * @param {object} email event object
+   * @returns {undefined} returns nothing
+   */
+  validate = (email) => {
```
I had previously said something about not using arrow functions in jsx context, pls change this and bind to the constructor if neccessary

rhenshaw56 commented on src/views/user/ResetPassword/ForgotPassword.jsx:
```diff
@@ -0,0 +1,153 @@
+    const emailErrors = [];
+    if (!email) emailErrors.push('Email cannot be empty');
+
+    const regex = new RegExp(['^(([^<>()[\\]\\.,;:\\s@"]+(\\.',
```
can you rename the regex variable name to be somewhat less generic ?

rhenshaw56 commented on src/views/user/ResetPassword/ForgotPassword.jsx:
```diff
@@ -0,0 +1,153 @@
+   * @param {object} errorResonse error message object
+   * @returns {string} error message
+   */
+  serverErrorResponse = (errorResonse) => {
```
please change all arrow functions to class methods

mppdg replied:
I have changed all arrow functions to class methods except for this and two others because Eslint flagged that they must have a `this` call on them if changed to the normal method. But they don't have any reference to the `the` instance.

rhenshaw56 commented on src/views/user/ResetPassword/ForgotPassword.jsx:
```diff
@@ -0,0 +1,153 @@
+}
+
+
+export const mapStateToProps = (state) => {
```
```
export const mapStateToProps = ({
  resetPasswordLink: { isLoading, success, error },
}) => ({
 isLoading,
 success,
 backendError: error
});
```

rhenshaw56 replied:
what do you think ?

mppdg replied:
It's robust!

rhenshaw56 commented on src/views/user/ResetPassword/ResetPassword.jsx:
```diff
@@ -0,0 +1,307 @@
+  componentDidMount() {
+    const { location, dispatch, history } = this.props;
+    const resetToken = location.search.substr(7);
+    axiosInstance().get(`/users/reset-password?token=${resetToken}`)
```
I won't allow you to do this in the context of a react component, pls move this to an action to update state, that's what actions are for, cos if this call errors, how are we going to set an error state in redux store

rhenshaw56 commented on src/views/user/ResetPassword/ResetPassword.jsx:
```diff
@@ -0,0 +1,307 @@
+      this.validity.passwordIsValid = false;
+      return;
+    }
+    const regex = new RegExp(['^(?=.*[\\d])(?=.*[A-Z])(?=.*[a-z])',
```
renname regex

rhenshaw56 replied:
so I see you having plenty regex everywhere, do you think you can have a utils folder where you have helpers handling all regex cases, instead of doing that in a component

mppdg replied:
I used them only in two instances and both cases handle different patterns. So, they are not reused.

rhenshaw56 commented on src/views/user/ResetPassword/ResetPassword.jsx:
```diff
@@ -0,0 +1,307 @@
+   * @returns {undefined} undefined
+   */
+  handlePasswordVisibility = (event) => {
+    const el = event.target;
```
what is `el` ?

mppdg replied:
It's the target element. The element that triggered the event. I have renamed it to `target`.

mppdg committed:
Merge branch 'staging' of https://github.com/andela/rivendell-ah-client into staging

rhenshaw56 commented on src/views/user/ResetPassword/ForgotPassword.jsx:
```diff
@@ -0,0 +1,164 @@
+    const serverResponse = { message: '', style: '' };
+
+    if (backendError.status) {
+      serverResponse.message = this.serverErrorResponse(backendError);
```
`I have changed all arrow functions to class methods except for this and two others because Eslint flagged that they must have a this call on them if changed to the normal method. But they don't have any reference to the the instance` 

but now it's on the instance itself,
so either you move it to another file or you pull it out of the context of the component entirely

mppdg replied:
I have moved it to another file

rhenshaw56 commented on src/views/user/ResetPassword/ForgotPassword.jsx:
```diff
@@ -0,0 +1,164 @@
+  dispatch: PropTypes.func.isRequired,
+};
+
+export default connect(mapStateToProps, null)(ForgotPassword);
```
what happens if there is no null ?

mppdg replied:
It doesn't actually change anything. I thought it was required. I have removed it. Thanks

rhenshaw56 commented on src/views/user/ResetPassword/ResetPassword.jsx:
```diff
@@ -0,0 +1,340 @@
+   * @param {Object} errorResonse http error status code
+   * @returns {String} error response message
+   */
+  serverErrorResponse = (errorResonse) => {
```
👀

mppdg replied:
It's now been moved to a helper file

mppdg committed:
feat(reset-password) user should be able to reset password

- add forget password links to login and signup page
- create a page where user can enter email
- ensure user receive an email
- redirect user to reset password page
- update user password
[Delivers #160056373]
